--- a/include/openssl/ssl.h.in	2023-01-13 14:54:07.824939000 +0100
+++ b/include/openssl/ssl.h.in	2023-01-13 15:32:24.232884727 +0100
@@ -195,9 +195,11 @@
  * DEPRECATED IN 3.0.0, in favor of OSSL_default_ciphersuites()
  * Update both macro and function simultaneously
  */
-#  define TLS_DEFAULT_CIPHERSUITES "TLS_AES_256_GCM_SHA384:" \
-                                   "TLS_CHACHA20_POLY1305_SHA256:" \
-                                   "TLS_AES_128_GCM_SHA256"
+#define TLS_DEFAULT_CIPHERSUITES    "TLS_AES_256_GCM_SHA384:" \
+                                    "TLS_CHACHA20_POLY1305_SHA256:" \
+                                    "TLS_AES_128_CCM_SHA256:" \
+                                    "TLS_AES_128_CCM_8_SHA256:"\
+                                    "TLS_AES_128_GCM_SHA256"
 # endif
 /*
  * As of OpenSSL 1.0.0, ssl_create_cipher_list() in ssl/ssl_ciph.c always
